backend/tracim_backend/views/core_api/session_controller.py

```diff
@@ -26,21 +30,36 @@ class SessionController(Controller):
     @hapic.input_headers(LoginOutputHeaders())
     @hapic.input_body(BasicAuthSchema())
     @hapic.output_body(UserSchema())
-    def login(self, context, request: TracimRequest, hapic_data=None):
+    def login(self, context, request: TracimRequest, hapic_data: HapicData) -> UserInContext:
         """
         Logs the user into the system.
         In case of success, the JSON returned is the user profile.
         In that case, a cookie is created with a session_key and an expiration date.
         Eg. : `session_key=932d2ad68f3a094c2d4da563ccb921e6479729f5b5f707eba91d4194979df20831be48a0; expires=Mon, 22-Oct-2018 19:37:02 GMT; Path=/; SameSite=Lax`
         """
 
-        login = hapic_data.body
+        login = hapic_data.body  # type: LoginCredentials
         app_config = request.registry.settings["CFG"]  # type: CFG
         uapi = UserApi(None, session=request.dbsession, config=app_config)
         ldap_connector = None
         if AuthType.LDAP in app_config.AUTH_TYPES:
             ldap_connector = get_ldap_connector(request)
-        user = uapi.authenticate(login.email, login.password, ldap_connector)
+
+        user = None
+        if login.email:
+            try:
+                user = uapi.authenticate(
+                    login=login.email, password=login.password, ldap_connector=ldap_connector,
+                )
+            except AuthenticationFailed as exc:
+                if not login.username:
+                    raise exc
+
+        if user is None:
+            user = uapi.authenticate(
+                login=login.username, password=login.password, ldap_connector=ldap_connector,
+            )
+
```

Comment:
It seems this could be refactored. the second authentication does not seem to be checked for exceptions, too.

```suggestion

        user = None
        try:
            user = uapi.authenticate(
                login=login.email if login.email else login.username,
                password=login.password,
                ldap_connector=ldap_connector,
            )
        except AuthenticationFailed as exc:
            if not login.username:
                raise exc
```

But the ternary `login.email if login.email else login.username` should probably be put in a dedicated method of `login`.

Reply:
Also, why is the execption only raised if `not login.username`?

Reply:
In your proposition, if `email` and `username` given, authentication is only tried with email (maybe username could work). 

> Also, why is the execption only raised if not login.username?

For the same reason, if email and username given, authentication should be tried with both of them.

Reply:
Oh, right, I hadn't thought that both the username and the email could be given. Is there a use case for this?

If so, what about:

```suggestion

        user = None
        for l in (login.email, login.username):
            try:
                user = uapi.authenticate(
                    login=l, password=login.password, ldap_connector=ldap_connector,
                )
            except AuthenticationFailed as exc:
                raise exc

            if user:
                break
```

(not very fond of the `break` here though)

Reply:
Because in the current state of the code , if the login.email is here, authentication will be tried, if it fails, an exception will be raised and the username-based authentication will not be tried, present or not, if I understand the code correctly.

Reply:
> Oh, right, I hadn't thought that both the username and the email could be given. Is there a use case for this?

Ii think it can be. When client don't know if user give username or email.

In [this](https://github.com/tracim/tracim/pull/2957#discussion_r419308828) proposition, if email and username given, and authentication with email fail, exception raise before try with username too.

> Because in the current state of the code , if the login.email is here, authentication will be tried, if it fails, an exception will be raised and the username-based authentication will not be tried, present or not, if I understand the code correctly.

Maybe i don't understand the current code correctly because what i understand is "if the login.email is here, authentication will be tried, if it fails, an exception will be raised **only if no username given**".

Reply:
right, but then shouldn't the second authentication raise in case of failure?

Reply:
Ok, I'm dumb, uapi.authenticate WILL actually raise on failure. The original code is right.